shanghai/core_plugins/channel.py

```diff
@@ -55,6 +56,78 @@ def __init__(self, channel: Channel, user: User, info: dict=None):
         self.info = info if info is not None else {}
 
 
+class ChannelContext(ShadowAttributesMixin):
+
+    def __init__(self, network_context: NetworkContext, message: 'BaseMessage',
+                 *, logger: Logger=None):
+        self.network_context = network_context
+        self.network = network_context.network
+        self._message = message
+
+        if logger is None:
+            logger = get_logger('channel', f'{self._message.channel}@{self.network.name}',
+                                self.network.config)
+        self.logger = logger
+
+    def say(self, message):
+        self.network_context.send_cmd('PRIVMSG', self._message.channel, message)
+
+    def msg(self, target, message):
+        self.network_context.send_cmd('PRIVMSG', target, message)
+
+    @property
+    def members(self):
+        member_list = []
+        for lkey, joinobj in self.network_context.joins.items():
+            if self.network_context.chan_eq(lkey[0], self._message.channel):
+                member_list.append(joinobj.user)
+        return member_list
+
+
+class ChannelEventDispatcher(EventDispatcher):
+
+    async def dispatch(self, context, msg):
+        return await super().dispatch(msg.command, context, msg)
+
+
+class BaseMessage(Message):
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        # some convenience accessors
+        self.channel = self.params[0]
+        self.line = self.params[-1]
+        self.words = self.line.split()
+        self.sender = self.prefix.name
+
+    @classmethod
+    def from_message(cls, message: Message):
+        # command name is "ChannelMessage" etc.
+        return cls(cls.__name__, prefix=message.prefix, params=message.params,
+                   tags=message.tags, raw_line=message.raw_line)
+
+
+class ChannelMessage(BaseMessage):
+    def __repr__(self):
+        return f'<{self.__class__.__name__} type={self.command!r} ' \
+               f'sender={self.prefix.name!r} source={self.params[0]!r} ' \
+               f'message={self.params[-1]!r}>'
+
+
+class PrivateMessage(BaseMessage):
+    def __repr__(self):
+        return f'<{self.__class__.__name__} type={self.command!r} ' \
+               f'sender={self.prefix.name!r} message={self.params[-1]!r}>'
+
+
+class ChannelNotice(ChannelMessage):
+    pass
+
+
+class PrivateNotice(PrivateMessage):
```

Comment:
I always thought of "Query" as something you show the messages with someone in, i.e. a window. I can see how one could see it as a message in itself, though. Imo "PrivateMessage" is more appropriate.